### PR TITLE
Adding functionality to enforce key uniqueness for labels

### PR DESCRIPTION
**Pull request by kamaleshnneerasa**
## Description
**Enforcing Key uniqueness for labels:** Added support for checking duplicate keys while creating or updated labels. Added unit tests to verify correctness.

### Testing
Please describe the tests that you ran to verify your changes. Please summarize what did you test and what needs to be tested e.g. deployed and tested helm chart locally. 

### Checklist:
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been merged and published in downstream modules

### Documentation
Make sure that you have documented corresponding changes in this repository or [hypertrace docs repo](https://github.com/hypertrace/hypertrace-docs-website) if required.

### CODE REVIEW

File: labels-config-service-impl/src/main/java/org/hypertrace/label/config/service/LabelsConfigServiceImpl.java [hunk 1]
```diff
@@ -83,7 +83,8 @@ public void createLabel(
     try {
       RequestContext requestContext = RequestContext.CURRENT.get();
       CreateLabel createLabel = request.getLabel();
-      if (systemLabelsKeyLabelMap.containsKey(createLabel.getKey())) {
+      if (systemLabelsKeyLabelMap.containsKey(createLabel.getKey())
```
Comment (saxenakshitiz) [inline labels-config-service-impl/src/main/java/org/hypertrace/label/config/service/LabelsConfigServiceImpl.java:86]: You need synchronisation across these multiple calls. Refer to https://github.com/hypertrace/config-service/blob/1a5f85c34f62ce96b901ef11f16b36c1a23f44b6/config-service-impl/src/main/java/org/hypertrace/config/service/store/DocumentConfigStore.java#L45 for syncronization across critical section. Even this doesn't work on multi-pod deployment. However we can start with this.

File: labels-config-service-impl/src/main/java/org/hypertrace/label/config/service/LabelsConfigServiceImpl.java [hunk 2]
```diff
@@ -80,32 +85,41 @@ private static Label buildLabelFromConfig(ConfigObject configObject) {
...
+    RequestContext requestContext = RequestContext.CURRENT.get();
+    Lock labelsLock = this.stripedLabelsLock.get(requestContext.getTenantId());
     try {
-      RequestContext requestContext = RequestContext.CURRENT.get();
-      CreateLabel createLabel = request.getLabel();
-      if (systemLabelsKeyLabelMap.containsKey(createLabel.getKey())) {
-        // Creating a label with a name that clashes with one of system labels name
-        responseObserver.onError(new StatusRuntimeException(Status.ALREADY_EXISTS));
-        return;
+      boolean lockAcquired = labelsLock.tryLock(5, TimeUnit.MINUTES);
```
Comment (saxenakshitiz) [inline labels-config-service-impl/src/main/java/org/hypertrace/label/config/service/LabelsConfigServiceImpl.java:91]: Please define a static final variable of type Duration, and then use it everywhere. Also `5 seconds` is max you should wait.

File: labels-config-service-impl/src/main/java/org/hypertrace/label/config/service/LabelsConfigServiceImpl.java [hunk 3]
```diff
@@ -38,6 +41,8 @@
   private final List<Label> systemLabels;
   private final Map<String, Label> systemLabelsIdLabelMap;
   private final Map<String, Label> systemLabelsKeyLabelMap;
+  private final int LABEL_LOCK_STRIPE_COUNT = 1000;
```
Comment (saxenakshitiz) [inline labels-config-service-impl/src/main/java/org/hypertrace/label/config/service/LabelsConfigServiceImpl.java:44]: static final

File: labels-config-service-impl/src/main/java/org/hypertrace/label/config/service/LabelsConfigServiceImpl.java [hunk 4]
```diff
@@ -38,6 +41,8 @@
...
+  private final Striped<Lock> stripedLabelsLock = Striped.lazyWeakLock(LABEL_LOCK_STRIPE_COUNT);
```
Comment (saxenakshitiz) [inline labels-config-service-impl/src/main/java/org/hypertrace/label/config/service/LabelsConfigServiceImpl.java:45]: Initialize in constructor

File: labels-config-service-impl/src/main/java/org/hypertrace/label/config/service/LabelsConfigServiceImpl.java [hunk 5]
```diff
@@ -133,41 +147,71 @@ public void getLabels(
...
+    } finally {
+      labelsLock.unlock();
     }
   }
+
+  private boolean isDuplicateKey(String key) {
+    RequestContext requestContext = RequestContext.CURRENT.get();
+    List<Label> labelList = configServiceCoordinator.getAllLabels(requestContext);
+    List<String> keyList =
+        labelList.stream().map(label -> label.getKey()).collect(Collectors.toList());
```
Comment (saxenakshitiz) [inline labels-config-service-impl/src/main/java/org/hypertrace/label/config/service/LabelsConfigServiceImpl.java:214]: You can `filter` here and use `findAny` to break this loop early

File: labels-config-service-impl/src/main/java/org/hypertrace/label/config/service/LabelsConfigServiceImpl.java (hunk 2 above)
Comment (kamaleshnneerasa) [reply labels-config-service-impl/src/main/java/org/hypertrace/label/config/service/LabelsConfigServiceImpl.java]: Done.

File: labels-config-service-impl/src/main/java/org/hypertrace/label/config/service/LabelsConfigServiceImpl.java [hunk 6]
```diff
@@ -80,32 +89,42 @@ private static Label buildLabelFromConfig(ConfigObject configObject) {
...
+        Label label =
+            Label.newBuilder()
+                .setId(UUID.randomUUID().toString())
+                .setKey(createLabel.getKey())
+                .build();
+        Label createdLabel = configServiceCoordinator.upsertLabel(requestContext, label);
+        responseObserver.onNext(CreateLabelResponse.newBuilder().setLabel(createdLabel).build());
+        responseObserver.onCompleted();
+      } else {
+        throw new StatusRuntimeException(Status.DEADLINE_EXCEEDED);
```
Comment (saxenakshitiz) [inline labels-config-service-impl/src/main/java/org/hypertrace/label/config/service/LabelsConfigServiceImpl.java:114]: Can we change it to line to -
```
responseObserver.onError(new StatusRuntimeException(Status.ABORTED));
return;
```
Change in other places as well

File: labels-config-service-impl/src/main/java/org/hypertrace/label/config/service/LabelsConfigServiceImpl.java [hunk 7]
```diff
@@ -80,32 +89,42 @@ private static Label buildLabelFromConfig(ConfigObject configObject) {
...
+    boolean lockAcquired = false;
     try {
-      RequestContext requestContext = RequestContext.CURRENT.get();
-      CreateLabel createLabel = request.getLabel();
-      if (systemLabelsKeyLabelMap.containsKey(createLabel.getKey())) {
-        // Creating a label with a name that clashes with one of system labels name
-        responseObserver.onError(new StatusRuntimeException(Status.ALREADY_EXISTS));
-        return;
+      lockAcquired = labelsLock.tryLock(WAIT_TIME.getSeconds(), TimeUnit.SECONDS);
+      if (lockAcquired) {
```
Comment (saxenakshitiz) [inline labels-config-service-impl/src/main/java/org/hypertrace/label/config/service/LabelsConfigServiceImpl.java:97]: Remove variable `lockAcquired`. Do as below -
```
if (labelsLock.tryLock(WAIT_TIME.getSeconds(), TimeUnit.SECONDS))
```

File: labels-config-service-impl/src/main/java/org/hypertrace/label/config/service/LabelsConfigServiceImpl.java [hunk 8]
```diff
@@ -80,32 +89,42 @@ private static Label buildLabelFromConfig(ConfigObject configObject) {
...
+        CreateLabel createLabel = request.getLabel();
```
[the remark this answers is not in the document]
Comment (kamaleshnneerasa) [reply labels-config-service-impl/src/main/java/org/hypertrace/label/config/service/LabelsConfigServiceImpl.java:98]: I don't think this is possible. if (labelsLock.tryLock(WAIT_TIME.getSeconds(), TimeUnit.SECONDS)) should be in a try block.

**Review by saxenakshitiz — APPROVED** [review]
lgtm

**Pull request merged**